### ISSUE TITLE
Increase the maximum number of dynamic rules to 30,000

### ISSUE THREAD

**Issue by 105th** [hober/web-extensions#319]
In the case of ad blockers, dynamic rules can serve several purposes:

1. Differential updates for the static rulesets.
2. Allowing users to add custom filter lists which are not included into the extension packages.
3. Applying custom rules made by the users themselves.

## Differential updates

The limit of 5,000 is simply not enough for this purpose. The problem is pretty much the same as with the current [“disableRules” limit](https://github.com/w3c/webextensions/issues/162#issuecomment-1309390759). Here are some graphs that show how many rules are being added to popular filter lists that are used by AdGuard users.

![Number of added network rules to AdGuard popular global lists](https://user-images.githubusercontent.com/11395678/200950749-7f65d434-5731-4781-9649-c530b9dc9d45.svg)

[Raw data](https://docs.google.com/spreadsheets/d/1MLdH_R6M_wqk3uCqHqqUvVSk0xWvN7twsa4R69teTuc/edit?usp=sharing)

On average about **2600+** new rules are added every week to the most popular filter lists. Note, that this does not include changes to regional lists (many users have 1-3 regional lists). Considering the release cycle of 3-6 weeks, we need to increase the dynamic rules limit to at least **20,000** just to make it possible to use it for differential updates.

## Custom filter lists

Ad blockers provide users with the possibility to add custom filter lists, the ones that are not included into the list of static rulesets. This is crucial for the ecosystem, that’s how new filter lists appear and become popular and eventually get included into the static rulesets. Is 5,000 enough for this purpose?

We analyzed public issues from AdGuard users from the [AdguardFilters repo](https://github.com/AdguardTeam/AdguardFilters/issues) to see what sorts of custom lists are used and how many rules are there on average. It appears that about 5% of the reports come from the users that have custom lists. And every one in four of those users uses more than the maximum limit of 5,000 rules.

## Custom user-made rules

Unfortunately, there’s no open stats that we can share on that. Although, from our experience, users rarely have over several hundreds of user-made rules.

## Summary

Considering all the cases that are listed above we suppose that the adequate limit of dynamic rules is **30,000**.

**Comment by Rob--W** [comment]
As noted in the meeting (https://github.com/w3c/webextensions/blob/main/_minutes/2022-12-08-wecg.md), this would also be a good time to separate the rule limits for session and dynamic rules.

A potential downside of separating the quotas for session and dynamic rules is that extensions could try to maximize the number of rules by reading and installing session rules at startup. In my opinion, if there is a concern on the registering session rules at startup, then the more sensible thing to do is to have lower limits for session rules, than to continue maintaining a shared limit for session + dynamic rules. E.g. continue limiting session rules to 5k, and expanding dynamic session rules to 30k.

**Comment by ameshkov** [comment]
@oliverdunk hi Oliver, we're waiting for some feedback from your team on this one, any chance you had an internal discussion?

**Comment by oliverdunk** [comment]
> @oliverdunk hi Oliver, we're waiting for some feedback from your team on this one, any chance you had an internal discussion?

Hi Andrey 👋  Nothing specific to share yet, but we have been discussing this and I'm personally very keen to address it.

If you're up for gathering some more data, one thing I would be interested in is the breakdown of the types of rules you'd be dynamically adding (how many are block rules, how many are redirects to extension URLs, how many are external redirects, how many are modifications etc).

That would be really useful in figuring out what sort of usage to expect if we were to increase this limit :)

**Comment by ameshkov** [comment]
Got it, we'll update the stats then, thank you!

**Comment by 105th** [comment]
@oliverdunk hi Oliver, we collected needed data with categories.

Filters were grouped by [the same](https://docs.google.com/spreadsheets/d/1MLdH_R6M_wqk3uCqHqqUvVSk0xWvN7twsa4R69teTuc/edit?usp=sharing) 3 groups: popular, recommended and others - they are represented on three pictures, one picture for one group of filters.

Added network rules divided into 6 categories:
- `added_allow` - allow rules with ['@@'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#basic-rules-syntax), which 'cancel' another rules,
- `added_allow_all_requests` - [`@@$document`](https://adguard.com/kb/general/ad-filtering/create-own-filters/#document-modifier),
- `added_redirect` - ['$redirect'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#redirect-modifier), ['$removeparam'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#removeparam-modifier), ['$redirect-rule'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#redirect-rule-modifier),
- `added_modify_headers` - ['$removeheader'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#removeheader-modifier), ['$csp'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#csp-modifier), ['$cookie'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#cookie-modifier),
- `added_badfilter` - ['$badfilter'](https://adguard.com/kb/general/ad-filtering/create-own-filters/#badfilter-modifier),
- `added_block` - all other;

The top chart shows all categories of rules, the bottom chart shows everything except `block` rules.

All these categories were extracted from [DNR RuleActionType](https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/#type-RuleActionType) and from [our process of conversion](https://github.com/AdguardTeam/tsurlfilter/blob/master/packages/tsurlfilter/src/rules/declarative-converter/grouped-rules-converters/abstract-rule-converter.ts#L349).

You can see that block rules take up about **98-99% of the changes**.

**Popular filters**
<img width="1484" alt="popular" src="https://github.com/w3c/webextensions/assets/11395678/cf2dbe1f-aa75-4b43-ba58-78a9dd324e9c">

**Recommended filters**
<img width="1477" alt="recommended" src="https://github.com/w3c/webextensions/assets/11395678/19808ebb-f343-49d9-82fc-365bf2bf144c">

**Other filters**
<img width="1476" alt="other" src="https://github.com/w3c/webextensions/assets/11395678/0e1de8e0-646f-491b-8fb9-dfa53b6791ea">


[Raw data here](https://docs.google.com/spreadsheets/d/13Nj3XwTU5D7YOCWyKXGAYoTOqCbwKL3se7KWv2GBa-4/edit?usp=sharing)

**Comment by oliverdunk** [comment]
@105th I forget if I said already but thank you so much for this! We get a lot from all of the data you share and really appreciate it - will try to have some follow-up on this for the next meeting :)

**Comment by oliverdunk** [comment]
We've been discussing this request for some time on the Chrome team and have put together a new proposal:

[**Proposal: Safe rules in declarativeNetRequest**](https://docs.google.com/document/d/1srkkCJkl4X2KOOUwnpDd-kvm8AY70eZBAYI4m836bvQ/edit?usp=sharing)

In short, the idea is to establish a smaller subset of rule types which are subject to an increased dynamic rule limit. There is a lot more detail in the document, so please do take a look. Importantly, there is no change to what you can do with rules in declarativeNetRequest or the existing limits - this is purely additive.

Feedback is very much welcome - in particular, the "safe rules" naming is something we'd like to discuss (no rule is definitively safe, just as all rules have valid use cases). We also want to make sure this addresses the most important use cases and fits in to the bigger picture of how the declarativeNetRequest API is used (for example, I know supporting differential updates is a key goal that may require more than one moving piece to fully support).

Happy to discuss here for now although we could consider opening a new issue if that would work better.

**Comment by Rob--W** [comment]
@oliverdunk Your proposal document currently proposes a higher limit for "block", "allow" and "allowAllRequests". Can it be just "allow" and "block", without "allowAllRequests"?

"allowAllRequests" is expensive in its design and it is not clear whether the request here actually needs it.

**Comment by oliverdunk** [comment]
> @oliverdunk Your proposal document currently proposes a higher limit for "block", "allow" and "allowAllRequests". Can it be just "allow" and "block", without "allowAllRequests"?
> 
> "allowAllRequests" is expensive in its design and it is not clear whether the request here actually needs it.

Assuming there aren't any use cases we're missing, I don't have any strong feelings there. I added it for completeness but "block" and "allow" are definitely the important ones.

**Comment by lyphyser** [comment]
Just remove the limit on the number of rules. There is of course no need to have a limit and an arbitrary magic number is obviously always wrong since there is no way to prove that the choice is optimal or correct. 

The limit is suspected to have been added by Google for anti-user purposes (enriching their ad business at the expense of the user by making ad blockers less effective), or was possibly added due to sheer incompetence.

If you want to prevent DoS by high memory usage, then account the rules as memory usage by the extension, and act on the memory usage of the whole extensions. Obviously this must not be an arbitrary limit, but instead some way of informing the user that the extension might be using too much memory and might be causing performance or crash issues.

(in case it's not obvious, with a proper implementation (i.e. a trie for simple patterns or NFA for regexes) performance does not depend asymptotically on the number of rules, but only on the size of the text to match, excluding the construction of the data structure which is irrelevant since it can be done in the background before the extension update is finalized and saved to disk)

**Comment by ameshkov** [comment]
@oliverdunk 

First of all, it's awesome that you agree to increase the dynamic limit, thank you!

Generally, the proposal covers what was requested pretty well.

The concept of "unsafe rules" seems interesting to me, maybe it can be used further in the future (optional permission?). Just one question: why `upgradeScheme` is not considered safe in the proposal?

**Comment by oliverdunk** [comment]
> @oliverdunk
> 
> First of all, it's awesome that you agree to increase the dynamic limit, thank you!
> 
> Generally, the proposal covers what was requested pretty well.
> 
> The concept of "unsafe rules" seems interesting to me, maybe it can be used further in the future (optional permission?). Just one question: why `upgradeScheme` is not considered safe in the proposal?

Thanks for taking a look at this!

I don't think there's any specific reason to exclude `upgradeScheme` - I'd be curious if you have thoughts on the use cases for this or how common you think it would be.

The main reason to focus on `block`, `allow` and `allowAllRequests` for now was just to keep this as well scoped as possible, but we're definitely open to conversations if it seems like what we have is too restrictive.

**Comment by ameshkov** [comment]
`upgradeScheme` is not too useful for us specifically since it's more for extensions like HTTPS everywhere, but I assume they wouldn't be against increasing the dynamic limit either. I am just trying to think of any scenario where `upgradeScheme` can be problematic and so far can't think of any.

**Comment by sammacbeth** [comment]
To give some context on `upgradeScheme` rules, in the context of DuckDuckGo's Smarter Encryption feature (which is what HTTPS everywhere is based off). We currently ship a list of ~250k domains to be upgraded in the MV3 version of our extension. By using the `requestDomains` option, we can encode these into 10 DNR rules, so the rule cap is currently not an issue for this use-case. We ship these as static rules anyway due to the list size - the ruleset file is currently 4.7Mb.

I wrote up some notes on the MV3 implementation for this feature, in case it is of interest: https://sammacbeth.eu/mv3-smarter-encryption

**Comment by SHAGGAR** [comment]
Has any consideration been given to letting the user choose limits for each extension?  

It seems like limiting the maximum rule count for safety reasons doesn't make any sense.  All it would take is one malicious rule to do damage.  Maybe the user can take on the accepted (theoretical) risk of large dynamic lists by choosing to enable and/or set a max value.  This would allow users who want to retain maximum functionality to eliminate the limits entirely, while the concern google et.al. have of legal risk for "approved" extensions goes away.  Alternatively just slap a warning on the extension page that says the filters are dynamic and may result in unexpected behavior.  We see this today for things like video games where a single player campaign is rated, but the online play is not because its out of control of the developer.  

TBH this whole thing seems pretty silly since the entire concept of a web browser is that you're going to view totally dynamic, unknown content every time you load a page.  I mean right now you could go to a search engine and get an ad that leads to a scam. Or maybe you load a page and a tracker builds a profile of you that it sells to advertisers/scammers.  Or maybe there's a vulnerability in the browser and a malicious website uses it to gain elevated access to your computer.

These are all risks we accept with the web and I don't think anyone from w3c or any of the browser vendors would ever consider tackling those issues in the form of a pre-approved browser whitelist.  So why are we doing that here with MV3?  

Let the user decide what risk they're willing to take.

**Comment by xeenon** [comment]
While the idea of allowing users to set their own limits for each extension is interesting, it introduces complexities, especially in terms of performance. The increase of the [content blocker extension rule limit in WebKit to 150,000](https://github.com/WebKit/WebKit/commit/d069434deaa0a8844ce6aa2090b0f61312f4bf45) was a balance between functionality and memory usage — see the linked commit for the juicy details. (That WebKit limit is also separate from Safari's dNR limits, for other technical reasons.)

Allowing users to choose their own limits could lead to performance issues in a way that might not be immediately obvious to them. This is crucial, considering the diverse range of devices and operating systems browsers support. Performance degradation due to high rule counts might be hard for users to diagnose, leading to a suboptimal browsing experience.

**Comment by SHAGGAR** [comment]
As a ublock origin user I have never had an issue with performance related to its dynamic lists.  Are these performance issues a new problem resulting from the design adopted for MV3?  If so perhaps that design should be reconsidered or abandoned.  

If this is not an issue of the MV3 design and the same issue exists in MV2 then why have users not noticed it yet?  Is it maybe because the performance hit of the blocker's lists is nothing compared to the performance hit of running the unblocked javascript of ads?  

I think most users would accept the insignificant performance hit of unlimited dynamic lists if it eliminates the extremely noticeable and unbearable performance cost of unblocked ads.  It seems like it would be better to let the user make that choice.

If the issue is low end devices not handling the lists, then maybe those devices should have limits set based on available resources.  Setting a static limit for all extensions across all hardware to support the lowest common denominator seems like not just a bad choice but also unnecessary. Browsers could automatically size lists to resources.

And even if there is an issue with low end devices we run into the same tradeoff: blocking the javascript of the ads is almost always going to be a vast performance improvement.  Effective ad blockers are even more critical for low end devices as any ads that sneak through will be detrimental to the user's experience.

**Comment by Dealman** [comment]
> While the idea of allowing users to set their own limits for each extension is interesting, it introduces complexities, especially in terms of performance. The increase of the [content blocker extension rule limit in WebKit to 150,000](https://github.com/WebKit/WebKit/commit/d069434deaa0a8844ce6aa2090b0f61312f4bf45) was a balance between functionality and memory usage — see the linked commit for the juicy details. (That WebKit limit is also separate from Safari's dNR limits, for other technical reasons.)
> 
> Allowing users to choose their own limits could lead to performance issues in a way that might not be immediately obvious to them. This is crucial, considering the diverse range of devices and operating systems browsers support. Performance degradation due to high rule counts might be hard for users to diagnose, leading to a suboptimal browsing experience.

The keyword here being _could_. Surely some diagnostics to alert the user if an extension is using a substantial amount of memory is a rather nice trade-off?

Looking at those numbers, I really don't see any issue other than for potentially low-end mobile devices. Whereas high-end devices will barely notice a few hundreds of MB being used.

Why resort to this kind of collective punishment only to satisfy low-end devices? I wholly agree with the other statements that some of the content that could otherwise be blocked but wouldn't, because of these imposed limits could have a far bigger performance impact.

While I like to try and keep things on-topic and professional, this rather obviously isn't decided on with the end-user in interest. That's but a façade, what's truly in interest here is Google and that alone.

I mean, W3C's own guidelines champion that the user should be able to render the content the way _they_ want - including assistive extensions and blocking of unwanted content. Why deny the user the ability to tweak these values themselves, if it means they could do just that?

The fact the limit was set to a shockingly low value of 5,000 just further cements that Google has no end-user interest in this. Even if increased to 30,000, that's laughably low when like you demonstrated - WebKit allows for 150,000.

I hate to sound a bit pessimistic, but I feel there's going to be some real Streisand effect fallout from all of this if it goes through the way Google wants it to.

**Comment by Rob--W** [comment]
FYI Firefox hasn't bumped the limit yet, it is tracked as part of https://bugzilla.mozilla.org/show_bug.cgi?id=1803370.

**Comment by xeenon** [comment]
This has been [fixed in WebKit](https://github.com/WebKit/WebKit/pull/26070), not shipping in Safari yet.

**Comment by imothep85** [comment]
I would like to add something VERY important to me here: 
my extension for chrome, firefox, opera, vivaldi, brave has been working for years and currently my extension blocks more than 1m dangerous and undesirable urls!

I'm redoing my extension because we all need to migrate to v3, otherwise V2 will disappear from chrome webstore extensions and others too.

I'm disgusted, angry and extremely nervous, because I learn that Google has set a limit of 5000/30000?

my extension blocks md5-encrypted urls txt file hosted on a server, and as I've already explained, I have a lot of them, and this database is updated according to my needs, and the needs of those who use my extension.

Is there a quick solution?
Extension must migrate to v3 absolutely, before disappearing from the stores before June 2025!  

my extension fetch a file from an url, that file is encrypted md5 urls, 1m of them.
i'm stuck with:  Failed to initialize blocking rules: Error: Dynamic rule count exceeded.
I tried TONS of things, nothing works !
I started from scratch, blocking a huge database doesn't work , local mode doesn't work ! 
I created a new extension for testing, EACH TIME the extension MV3 stop working RANDOMLY, 
the service worker switch between active and inactive .
In my case it block url stored in a local file, then suddenly it doesn't block the url from that file anymore !

https://github.com/user-attachments/assets/10b438ea-6534-480b-a689-22ea92005cd9

No message error, nothing in the console...
THE MV3 of Google is going to KILL every existing extension and i'm sure, that it's what they want to do, SHUTDOWN the chrome extensions store, preventing people to create new ones, and get a 100% maximum of forced ADS !
Thanks to everyone !

**Comment by jeltaqq** [comment]
@imothep85 
My guess is that Google will consider blocking everything and allowing 5000/30000.
I don't think we'll ever visit more than 5000 sites in our lifetime. (In terms of the number of unique websites (domains))

**Comment by Xevion** [comment]
> [@imothep85](https://github.com/imothep85) My guess is that Google will consider blocking everything and allowing 5000/30000. I don't think we'll ever visit more than 5000 sites in our lifetime. (In terms of the number of unique websites (domains))

That is, in no uncertain terms, utterly false. I wrote a little script to query the SQLite database of my browser, and tell me how many unique domains there were. Even added a little thing to remove any domains I deemed 'duplicates' (chrome extensions, IP addresses, CDNs, etc). and merged a bunch of them (removed about 1,000 domains). In just the last ~6 months, I've visited almost 4,000 sites.

```
> historee --top 5 --bottom 5 --redact

--- Vivaldi History Analysis ---
Date range: February 9, 2025 to August 20, 2025 (191 days)
Total unique domains found: 3,830
Domains removed (no valid TLD): 976

Top 5 most visited domains:
- *******.net: 15,927 visits
- ******.com: 12,893 visits
- ******.com: 7,160 visits
- ********.com: 5,059 visits
- ********.org: 4,997 visits

Bottom 5 least visited domains:
- ***.****.cc: 1 visit
- **********.com: 1 visit
- *******.com: 1 visit
- ******.******.com: 1 visit
- ********.app: 1 visit

> historee --all-browsers

--- All Browsers History Analysis ---
Date range: February 24, 2024 to August 20, 2025 (543 days)
Total unique domains found: 6,349
Domains removed (no valid TLD): 1,501
```

**Comment by ryanbr** [comment]
From Easylist, I pushed out a commit [6800 removal of dead/unused domains](https://github.com/easylist/easylist/commit/55fcc8d). Which would've caused all MV3 extensions relying on Easylist to break until another extension update. 

Not asking for a 30k dynamic limit, but if we can get the dynamic rule limit bumped to 10k or 12k/15k, just to avoid this from occurring in the future. 

Making wholesale large removals like this would've caused a lot of pain for extension authors, I'm wanting to avoid causing MV3  issues due to a large commit like this.

/Easylist Maintainer

**Comment by oliverdunk** [comment]
> From Easylist, I pushed out a commit [6800 removal of dead/unused domains](https://github.com/easylist/easylist/commit/55fcc8d). Which would've caused all MV3 extensions relying on Easylist to break until another extension update.

Great to hear from you! I appreciate you flagging this - I've been having a similar conversation with Eyeo recently. It sounds like these bulk removals aren't incredibly common but do happen a few times a year.

> Not asking for a 30k dynamic limit, but if we can get the dynamic rule limit bumped to 10k or 12k/15k, just to avoid this from occurring in the future.

At the moment, this isn't something we are prioritising. For context, we're not too concerned about the number of removals in general. However, extensions remove rules by adding them to a list of disabled rules, and that is implemented by storing them in Chrome's preferences file. That is good because it makes them available early during browser startup, but has the tradeoff that disabled rules are stored alongside other Chrome preferences. With the current limit disabled rules can already take up about half of that file and we don't want to exceed that as it may start to impact Chrome performance in general.

As far as I understand it, large removals shouldn't actually _break_ extensions, but would just prevent them from doing another ruleset update dynamically without an upload to the Chrome Web Store. We have [expedited review](https://developer.chrome.com/docs/webstore/skip-review) and rule removals would qualify, so they should be able to publish an update through the store instantly without any delay.

**Comment by ryanbr** [comment]
If rule removal can be done without counting, that would be a good first step.  I don't want cause all the extensions to suddenly not update because I decided to roll out a large cleanup of rules. 

When I say "break", Not allowing the extension to do list update is a broken feature (esp if its broken site, but the end user has to wait for an extension update). Currently I've been staggering my cleanup commits with Eyeo updates, and limiting the amount I can remove to ensure they don't the hit the MV3 limit prematurely. 

Hypothetically, even if we did a minor bump to the dynamic limit to say 10k, would give some list author flexibility to be less concerned about extension limits, without checking each extension on how much tokens they have left. We are also making more daily commits then ever before (Presentation in Cyprus) 

Hoping there is a better way

**Comment by Yuki2718** [comment]
Seriously, can't all these be processed a bit quicker? Frankly speaking, it's as if no progress has been made for a year. I know that's not the case, but this is my very straight feeling as one of those who *actually be burdened* by MV3.

**Comment by oliverdunk** [comment]
> If rule removal can be done without counting, that would be a good first step. I don't want cause all the extensions to suddenly not update because I decided to roll out a large cleanup of rules.

Rule removals need to be counted unfortunately, for the reasons I mentioned. Our way of storing those is not free but we also don't want to change it as it may impact our ability to have them available quickly at browser startup.

> When I say "break", Not allowing the extension to do list update is a broken feature (esp if its broken site, but the end user has to wait for an extension update).

I wonder if longer term there could be a way to differentiate rules being removed for cleanup vs. high priority removals. I wouldn't expect there to be 5000 broken sites between extension updates. This problem arises because the cleanup is using up the limit that is available.

> We are also making more daily commits then ever before (Presentation in Cyprus)

Interesting - definitely happy to chat more in Cyprus. I see the concern but I'm not sure there is a compelling enough argument for changing the limit yet.

**Comment by gorhill** [comment]
> definitely happy to chat more in Cyprus

I won't be there but for what it's worth, uBO Lite (uBOL) does not make use of that capability, as a good chunk of filter list updates is outside DNR rules anyways -- it's simpler to just update the whole extension with all the DNR rules and content scripts all updated at once. At the moment I try to update the extension once a week and for uBOL specifically (can't speak for other blockers) trying to patch DNR rules in between wouldn't be worth the added complexity and wouldn't be worth updating only the DNR rules without the cosmetic- or content script-based filters.

**Comment by ryanbr** [comment]
> > When I say "break", Not allowing the extension to do list update is a broken feature (esp if its broken site, but the end user has to wait for an extension update).
> 
> I wonder if longer term there could be a way to differentiate rules being removed for cleanup vs. high priority removals. I wouldn't expect there to be 5000 broken sites between extension updates. This problem arises because the cleanup is using up the limit that is available.

Consider if there is a broken site for a user, but the extension needs to update because it hit its 5k limit prematurely. An update by an extension to fix said broken site could take a while, depending when the author notices. The scenario could be very real, given my 6800 removal commit stop many extensions from updating lists. We do rely on real time fixes to go out.

Just for reference; large changes specifically for [`easylist_adservers.txt`](https://github.com/easylist/easylist/blob/master/easylist/easylist_adservers.txt) that would take a large chunk of tokens in one commit.

|  Date           |  Total    |  Net        |  Added | Deleted | Commit | Message |
|  ------------ | ---------| --------- | -------- | --------- | ----------- | ------------------------------- |
| 2025-08-14 | 6886     | -6886     | +0        | -6886     | 55fcc8dc   | M: Cleanup |
| 2025-07-16 | 4644     | -4644     | +0        | -4644     | 0eff22c6   | M: Cleanup |
| 2024-10-17 | 4239     | -4155     | +42       | -4197     | 9e3e171a   | M: Address https://github.com/easylist/e  |
| 2024-08-12 | 3705     | -2955     | +375      | -3330     | 907eec81   | A: https://filmstories.co.uk/ |
| 2024-11-04 | 3659     | -3263     | +198      | -3461     | bb091ebc   | M: Cleanup |
| 2025-02-15 | 2752     | -2680     | +36       | -2716     | c141298d   | M: Cleanup |
| 2025-06-29 | 2558     | -2556     | +1        | -2557     | 248716a4   | M: Update |
| 2024-10-10 | 1950     | -1770     | +90       | -1860     | daa59684   | M: Cleanup |
| 2025-03-29 | 1882     | -1752     | +65       | -1817     | e1306853   | M: Update |
| 2025-05-07 | 1800     | -1754     | +23       | -1777     | a0d210cd   | M: Cleanup |
| 2025-09-17 | 1643     | -1643     | +0        | -1643     | a7280dc1   | M: Cleanup |
| 2023-04-07 | 1619     | -1543     | +38       | -1581     | 83ebff34   | M: Cleanup |
| 2024-01-23 | 1551     | +1551     | +1551     | -0        | 3c3b5965   | M: Cleanup |
| 2024-01-23 | 1531     | -1437     | +47       | -1484     | 0bdae98a   | M: Cleanup old domains |
| 2024-06-02 | 1526     | -1004     | +261      | -1265     | 5bed6b45   | M: Update |
| 2024-06-29 | 1502     | -1134     | +184      | -1318     | db82c3df   | M: Cleanup |

Trying to Sync [CNAME changes ](https://github.com/AdguardTeam/cname-trackers/tree/master/data). Some of these changes are large, predating MV3 enforcement. But trying to do this task now, you can see doesn't take to much to bump the mv3 extension tokens in a single change.

|  Date           |  Total    |  Net        |  Added | Deleted | Commit | Message |
|  ------------ | ---------| --------- | -------- | --------- | ----------- | ------------------------------- |
| 2024-07-03 | 36285    | +15547    | +25916    | -10369   | d2e22aa2   | Update cnames (#19449) |
| 2023-04-27 | 24085    | +4687     | +14386    | -9699     | 1283e067   | M: Update CNAME's |
| 2023-07-03 | 21673    | -133      | +10770    | -10903    | 7957d409   | M: Update CNAME's |
| 2024-07-03 | 15858    | +0        | +7929     | -7929     | 5f464e9f   | M: Address https://github.com/Adgua |
| 2024-07-03 | 13957    | -13957    | +0        | -13957   | 86552bc5   | M: Remove redundant rules |
| 2022-09-10 | 13100    | +4190     | +8645     | -4455      | 4fda3672   | M: Sync CNAME trackers |
| 2022-02-08 | 8103     | +315      | +4209     | -3894      | 80d12b81   | M: Update CNAME trackers |

> 
> > We are also making more daily commits then ever before (Presentation in Cyprus)
> 
> Interesting - definitely happy to chat more in Cyprus. I see the concern but I'm not sure there is a compelling enough argument for changing the limit yet.

Is a philosophical reason not to consider changing the limits? Is there harm bumping it. Compared to last year for example, We are committing more changes daily (8-10x) and also have to make more removals because of it (talk in cyprus about this topic). Without the removals in easylist_adservers would become bloated, which is why we do these cleanups.

**Comment by Yuki2718** [comment]
TBH it is currently barely working only because it happened to be (in practice) only ABP/AdBlock that actually utilizes dynamic rules, which allows us to pre-cordinate when to do these cleanup.

**Comment by ryanbr** [comment]
@oliverdunk 

If we make a change of 20k removals (considered dead/unused domains) in Easylist, will this impact any of the MV3 limits of the extensions?

**Comment by oliverdunk** [comment]
Hey @ryanbr - we have a limit of 5,000 disabled static rules. This means that if an extension tries to use the `updateStaticRules()` API to disable more than 5,000 rules (specifically rules that were bundled with the last update) it will fail. The extension should continue to work but the behavior will depend on the how the developer handles this - in the best case it might mean that further removals are ignored.

There are ways developers can work around this. For example, they can submit a new update to the Chrome Web Store with the rules removed, and as long as nothing else changed the update will be approved immediately without review.

However, as this might be a problem for at least _some_ extensions, I would recommend against making such a big change at once and instead break it up into a few smaller (< 5000 removals, with some buffer for regular removals) changes.

I do want to see if we can relax these limits in the future (I know we discussed that together last year) and it's still something on our radar.

**Comment by stephenhawk8054** [comment]
I'm really pissed with MV3/Safari content blocker APIs situations (cc @xeenon ) across the browsers right now, but I'll try to calm myself down.

1. Removing rules (just the act of removing, the most harmless act of content blockers) is extremely limited for content blockers that utilize dynamic rules

---> As a result the main blocklists (that are contributed and used by communities) are so so hesitant to clean the lists right now, despite that there are so many dead domains inside those lists currently. And this leads to 2 --->

2. When the lists cannot be cleaned often, they just grow bigger and bigger and could reach the limit of static rules anytime. Fortunately, in MV3 `||example1.com^` and `||example2.com^` can be combined to 1 `requestDomains` rule and so the number of static rules at least can be still in control of content blockers despite the blocklist sources just grow bigger. **HOWEVER** --->

3. For the blockers that use MV3 on Safari, in which Safari will convert to their content blockers APIs rules, `||example1.com^` and `||example2.com^` are just counted separately, despite content blockers trying to combine them into 1 `requestDomains` rule

---> MV3 content blockers on Safari just throw errors now because they exceed 150k rules. As a result, uBOL have to [disable all of its malware/phishing lists to make room for other lists](https://github.com/uBlockOrigin/uBOL-home/issues/628#issuecomment-4085814675) (uBOL before this had to disable the whole list of removing tracking parameters from the URLs due to the same limitations). Yes, uBOL users can't have **MALWARE/PHISHING** protection lists due to the inconsistent implementation of MV3 between browsers.

---> As a result, MV3 content blockers on Safari now need blocklist sources to clean the dead domains often (in which the communities have already done it for years before MV3), but now removing too much will cause issue as in point 1 above again.

---

Seriously, can the browsers talk to each other for ONCE? What do you guys want us users/content blockers communities to do more? Do you know that the websites now do more and more aggressive ads/trackings/phishing/malwares/anti-adblock? Browsers now just give more power for websites to do anti-user practices everyday while users are constantly taken away most of their rights to use the websites safely.

**Comment by ShivanKaul** [comment]
> However, as this might be a problem for at least some extensions, I would recommend against making such a big change at once and instead break it up into a few smaller (< 5000 removals, with some buffer for regular removals) changes.

Respectfully, Google should be making it easier for open-source maintainers to ship security/privacy/compatibility enhancements, not asking them to take on even more work to get around a Google limitation.

**Comment by Mungr3ss** [comment]
on Adguard mv2 i can add all the filters from https://github.com/yokoffing/filterlists
but on adguard mv3 i can only add a couple of the smaller ones because the large ones go over the 30,000 limit.

can someone explain why mv2 allows 100's of thousands of filters but mv3 is limited to a pitiful 30,000 ? effectively making many filter lists useless on ALL chrome browsers.

firefox has no such limits. 

it's this kind of bs that makes switching to any of the chrome browsers a big fat NOPE.

i was using edge for a while as i was quite happy with the performance but the mv3 change sucks, so i returned to firefox. however i'm trying to have less 3rd party apps installed as i can't be bothered to reinstall lots of things these days.

is there any word on the 30,000 limit being increased further so filter lists become useful again on chrome browsers?

**Comment by SHAGGAR** [comment]
> can someone explain why mv2 allows 100's of thousands of filters but mv3 is limited to a pitiful 30,000 ? effectively making many filter lists useless on ALL chrome browsers.
> 
The goal of the limits is specifically to make ad blockers less effective. Chrome and the upstream chromium project are owned by google which is an advertising company. Obviously an advertising company would prefer their ads not be blocked by their own products. Other chromium based browsers have not yet removed mv2 support or have incorporated ad blockers that support large dynamic lists into their core functionality. 

> i was using edge for a while as i was quite happy with the performance but the mv3 change sucks, so i returned to firefox. however i'm trying to have less 3rd party apps installed as i can't be bothered to reinstall lots of things these days.
> 

Edge still allows MV2 extensions (though you may need to enable the following policy: https://learn.microsoft.com/en-us/deployedge/microsoft-edge-policies/extensionmanifestv2availability). They may fully remove mv2 at some point in the future but there is no info on when that will be at this time.

**Comment by Mungr3ss** [comment]
Thanks for explaining it in simple terms. I wonder if Google has shot themselves in the foot on this change? I know there will be clueless users who are still unaware they can block ads but won't this make more users switch to non crappy browsers?

I'd switch to edge but there's one thing that is stopping me and that's it's side panel that can't be resized or moved. microsoft suck at listening to feedback / take aggggeeeesss to implement things these days. seems all the chrome browsers have a garbage side panel that is either too big or badly designed. sorry for going off topic...

i don't know anyone who likes ads. i wish we ad blockers for tv (i know it's not websites) but imagine something that detected an ad break and mute the audio or showed a slideshow of landscapes for the duration and then detected when the programme resumed and auto switched back....man can dream.
